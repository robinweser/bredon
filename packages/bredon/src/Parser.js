/* @flow */
import tokenize from 'tokenize-sync'
import {
  dimension,
  integer,
  float,
  important,
  operator,
  identifier,
  expression,
  functionExpression,
  url,
  stringLiteral,
  parenthesis,
  hexColor,
  keyword,
  separator,
  cssValue,
  multiValue,
  isSeparator
} from 'bredon-types'

import isValidUnit from './utils/isValidUnit'
import isValidHexadezimal from './utils/isValidHexadezimal'
import isKeyword from './utils/isKeyword'
import getQuote from './utils/getQuote'
import parseMultiValue from './utils/parseMultiValue'

import type { Token } from '../../../flowtypes/Token'
import type {
  MultiValue,
  CSSValue,
  Node,
  SimpleNode,
  DimensionNode,
  FloatNode,
  FunctionNode
} from '../../../flowtypes/AST'

const ruleMap = {
  important: /^[!]$/,
  quote: /^('|\\'|"|\\")$/,
  operator: /^(\+|-|\*|\/)$/,
  identifier: /^[a-z-]+$/i,
  number: /^\d+$/,
  url_chars: /^[&:=?]$/,
  percentage: /^[%]$/,
  floating_point: /^[.]$/,
  hex: /^[#]$/,
  whitespace: /^\s+$/,
  paren: /^[()]$/,
  comma: /^,+$/
}

export default class Parser {
  currentPosition: number
  currentToken: Token
  parenBalance: number
  scope: string
  tokens: Array<Token>

  getNextToken(position: number): ?Token {
    const nextPosition = this.currentPosition + position

    if (nextPosition < this.tokens.length) {
      return this.tokens[nextPosition]
    }
  }

  updateCurrentToken(increment?: number = 0): void {
    this.currentPosition += increment
    this.currentToken = this.tokens[this.currentPosition]
  }

  isRunning() {
    return this.currentPosition < this.tokens.length
  }

  setScope(scope?: string): void {
    this.scope = scope || ''
  }

  walkTokens(): Node {
    this.updateCurrentToken()
    this.parseWhitespace()

    // skip parsing if no more tokens are coming
    // this happens if the strings ends with whitespace
    if (!this.currentToken) {
      return false
    }

    const node =
      this.parseOperator() ||
      this.parseHex() ||
      this.parseNumber() ||
      this.parseFloat() ||
      this.parseImportant() ||
      this.parseIdentifier() ||
      this.parseParenthesis() ||
      this.parseStringLiteral() ||
      this.parseSeparator()

    if (!node) {
      throw new SyntaxError(
        `Could not parse the token "${this.currentToken
          .type}" with the value "${this.currentToken.value}"`
      )
    }

    ++this.currentPosition
    return node
  }

  parse(input: string): MultiValue | CSSValue {
    this.tokens = tokenize(input, ruleMap)

    this.currentPosition = 0
    this.parenBalance = 0
    this.scope = ''

    const nodes = []

    while (this.isRunning()) {
      const node = this.walkTokens()

      if (node) {
        nodes.push(node)
      }
    }

    // split values into multi value if separator is present
    if (nodes.find(node => isSeparator(node) && node.value === ',')) {
      return multiValue(parseMultiValue(nodes))
    }

    return cssValue(nodes)
  }

  parseOperator(): SimpleNode {
    if (this.currentToken.type === 'operator') {
      if (this.scope === 'expression') {
        return operator(this.currentToken.value)
      }

      // parse value separator
      if (this.currentToken.value === '/') {
        return separator('/')
      }

      // parse numbers and floats with an algebraic sign
      if (this.currentToken.value === '+' || this.currentToken.value === '-') {
        const nextToken = this.getNextToken(1)

        if (nextToken) {
          if (nextToken.type === 'number') {
            const isNegative = this.currentToken.value === '-'

            this.updateCurrentToken(1)
            return this.parseNumber(isNegative)
          }

          if (nextToken.type === 'floating_point') {
            const isNegative = this.currentToken.value === '-'

            this.updateCurrentToken(1)
            return this.parseFloat(integer(0, isNegative))
          }

          throw new SyntaxError(
            `An addition or substraction sign must be used inside expressions ("calc()") or followed by a number/float. Instead found "${nextToken.value}" of type "${nextToken.type}".`
          )
        }

        throw new SyntaxError(
          'An addition or substraction sign must not be used at the end of a string.'
        )
      }

      throw new SyntaxError(
        'A multiplication sign (*) can only be used inside expression ("calc()").'
      )
    }
  }

  parseHex(): SimpleNode {
    if (this.currentToken.type === 'hex') {
      this.updateCurrentToken(1)
      let hexValue = ''

      while (
        this.isRunning() &&
        (this.currentToken.type === 'number' ||
          this.currentToken.type === 'identifier')
      ) {
        hexValue += this.currentToken.value
        this.updateCurrentToken(1)
      }

      const hexValueLength = hexValue.length
      // validating correct hex value length
      if (
        hexValueLength !== 3 &&
        hexValueLength !== 6 &&
        hexValueLength !== 8
      ) {
        // TODO: throw invalid hex value length
      }

      // validating correct hex value
      if (!isValidHexadezimal(hexValue)) {
        // TODO: throw invalid hex value
      }

      return hexColor(`#${hexValue}`)
    }
  }

  parseNumber(
    isNegative?: boolean = false
  ): SimpleNode | DimensionNode | FloatNode {
    if (this.currentToken.type === 'number') {
      const nextToken = this.getNextToken(1)

      // Parsing dimensions if a number is directly followed
      // by an identifier that matches any of the valid units
      if (
        nextToken &&
        (nextToken.type === 'identifier' || nextToken.type === 'percentage')
      ) {
        if (!isValidUnit(nextToken.value)) {
          throw new SyntaxError(
            `A number (${this.currentToken
              .value}) must be followed by a valid unit. Instead found "${nextToken.value}" of type "${nextToken.type}".`
          )
        }

        ++this.currentPosition

        return dimension(
          integer(parseInt(this.currentToken.value, 10), isNegative),
          nextToken.value.toLowerCase()
        )
      }

      const integerPart = integer(
        parseInt(this.currentToken.value, 10),
        isNegative
      )

      // Parsing floating point numbers if a number is directly followed
      // by a floating point which is followed by a number
      if (nextToken && nextToken.type === 'floating_point') {
        this.updateCurrentToken(1)
        return this.parseFloat(integerPart, isNegative)
      }

      return integerPart
    }
  }

  parseFloat(integerPart?: number): FloatNode {
    // floats can also start with a floating point
    // if the integer part is 0
    if (this.currentToken.type === 'floating_point') {
      const nextToken = this.getNextToken(1)

      if (nextToken) {
        if (nextToken.type === 'number') {
          this.updateCurrentToken(1)

          return float(
            integerPart || integer(0),
            integer(parseInt(nextToken.value, 10))
          )
        }

        throw new SyntaxError(
          `A floating point (.) must be followed by a number. Instead found ${nextToken.value} of type ${nextToken.type}.`
        )
      }

      throw new SyntaxError(
        'A floating point must not be used at the end of a string.'
      )
    }
  }

  parseImportant(): SimpleNode {
    if (this.currentToken.type === 'important') {
      const nextToken = this.getNextToken(1)

      if (nextToken) {
        if (
          nextToken.type === 'identifier' &&
          nextToken.value.match(/^important$/i) !== null
        ) {
          this.updateCurrentToken(1)
          return important()
        }

        throw new SyntaxError(
          `An exclamation mark (!) must always be followed by an "important" identifier. Instead found "${nextToken.value}" of type "${nextToken.type}".`
        )
      }

      throw new SyntaxError(
        'An exclamation mark (!) must not be used at the end of a string.'
      )
    }
  }

  parseIdentifier(): SimpleNode | FunctionNode {
    if (this.currentToken.type === 'identifier') {
      const nextToken = this.getNextToken(1)

      // parsing function expression
      if (nextToken && nextToken.type === 'paren' && nextToken.value === '(') {
        const callee = identifier(this.currentToken.value)

        // special parsing for url functions
        // as they only allow a single url parameter
        if (callee.value === 'url') {
          return this.parseURL(callee)
        }

        // special parsing for calc functions that includes
        // algebraic expressions with operators and parenthesis
        if (callee.value.indexOf('calc') !== -1) {
          // setting the expression scope
          // to allow all operators
          this.setScope('expression')
          const node = this.parseFunctionExpression(callee)
          this.setScope()

          const calcExpression = expression(node.params)
          node.params = [calcExpression]

          return node
        }

        return this.parseFunctionExpression(callee)
      }

      // parsing keywords into a special AST node
      if (isKeyword(this.currentToken.value)) {
        return keyword(this.currentToken.value)
      }

      return identifier(this.currentToken.value)
    }
  }

  parseFunctionExpression(callee: SimpleNode): FunctionNode {
    this.updateCurrentToken(2)
    ++this.parenBalance

    const startParenBalance = this.parenBalance

    const params = []

    while (
      this.isRunning() &&
      (this.currentToken.type !== 'paren' ||
        (this.currentToken.type === 'paren' &&
          this.currentToken.value !== ')') ||
        this.parenBalance !== startParenBalance)
    ) {
      // skip all commas directly
      while (this.isRunning() && this.currentToken.type === 'comma') {
        this.updateCurrentToken(1)
      }

      params.push(this.walkTokens())

      this.updateCurrentToken()
      this.parseWhitespace()
    }

    --this.parenBalance
    return functionExpression(callee, params)
  }

  parseURL(callee: SimpleNode): FunctionNode {
    let urlValue = ''

    this.updateCurrentToken(2)
    ++this.parenBalance

    const startParenBalance = this.parenBalance

    while (
      this.isRunning() &&
      (this.currentToken.type !== 'paren' ||
        (this.currentToken.type === 'paren' &&
          this.currentToken.value !== ')') ||
        this.parenBalance !== startParenBalance)
    ) {
      // concat any token to the url
      // no validation is done here!
      urlValue += this.currentToken.value
      this.updateCurrentToken(1)
    }

    --this.parenBalance
    return functionExpression(callee, [url(urlValue)])
  }

  parseStringLiteral(): SimpleNode {
    if (this.currentToken.type === 'quote') {
      const quote = getQuote(this.currentToken.value)

      let stringValue = ''
      this.updateCurrentToken(1)

      while (this.isRunning() && getQuote(this.currentToken.value) !== quote) {
        stringValue += this.currentToken.value
        this.updateCurrentToken(1)
      }

      return stringLiteral(stringValue, quote)
    }
  }

  parseParenthesis(): SimpleNode {
    if (this.currentToken.type === 'paren') {
      // update paren balance
      if (this.currentToken.value === '(') {
        ++this.parenBalance
      } else {
        --this.parenBalance
      }

      return parenthesis(this.currentToken.value)
    }
  }

  parseSeparator(): SimpleNode {
    if (this.currentToken.type === 'comma') {
      return separator(',')
    }
  }

  parseWhitespace() {
    if (this.currentToken.type === 'whitespace') {
      this.updateCurrentToken(1)
    }
  }
}
