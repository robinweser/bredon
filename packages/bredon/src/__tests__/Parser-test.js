import Parser from '../Parser'

describe('Parsing CSS values', () => {
  it('should correctly parse Identifiers', () => {
    const parser = new Parser()

    expect(parser.parse('flex-start')).toEqual({
      body: [
        {
          type: 'Identifier',
          value: 'flex-start'
        }
      ],
      type: 'CSSValue'
    })
  })

  it('should correctly parse Integers', () => {
    const parser = new Parser()

    expect(parser.parse('400')).toEqual({
      body: [
        {
          type: 'Integer',
          value: 400
        }
      ],
      type: 'CSSValue'
    })
  })

  it('should correctly parse algebraic signs', () => {
    const parser = new Parser()

    expect(parser.parse('-400')).toEqual({
      body: [
        {
          type: 'Integer',
          value: 400,
          negative: true
        }
      ],
      type: 'CSSValue'
    })
  })

  it('should correctly parse Parentheses', () => {
    const parser = new Parser()

    expect(parser.parse('(')).toEqual({
      body: [
        {
          type: 'Parenthesis',
          value: '('
        }
      ],
      type: 'CSSValue'
    })
  })

  it('should correctly parse HexColors', () => {
    const parser = new Parser()

    expect(parser.parse('#66FF66')).toEqual({
      body: [
        {
          type: 'HexColor',
          value: '#66FF66'
        }
      ],
      type: 'CSSValue'
    })
  })

  it('should correctly parse StringLiterals', () => {
    const parser = new Parser()

    expect(parser.parse('"hello, it\'s me."')).toEqual({
      body: [
        {
          type: 'StringLiteral',
          quote: '"',
          value: "hello, it's me."
        }
      ],
      type: 'CSSValue'
    })
  })

  it('should correctly parse Dimensions', () => {
    const parser = new Parser()

    expect(parser.parse('300px')).toEqual({
      body: [
        {
          type: 'Dimension',
          unit: 'px',
          value: {
            type: 'Integer',
            value: 300
          }
        }
      ],
      type: 'CSSValue'
    })
  })

  it('should correctly parse Float', () => {
    const parser = new Parser()

    expect(parser.parse('200.55')).toEqual({
      body: [
        {
          type: 'Float',
          fractional: {
            type: 'Integer',
            value: 55
          },
          integer: {
            type: 'Integer',
            value: 200
          }
        }
      ],
      type: 'CSSValue'
    })

    expect(parser.parse('-.55')).toEqual({
      body: [
        {
          type: 'Float',
          integer: {
            type: 'Integer',
            value: 0,
            negative: true
          },
          fractional: {
            type: 'Integer',
            value: 55
          }
        }
      ],
      type: 'CSSValue'
    })
  })

  it('should correctly parse FunctionExpressions', () => {
    const parser = new Parser()

    expect(parser.parse('rgba(200,300)')).toEqual({
      body: [
        {
          type: 'FunctionExpression',
          callee: {
            type: 'Identifier',
            value: 'rgba'
          },
          params: [
            {
              type: 'Integer',
              value: 200
            },
            {
              type: 'Integer',
              value: 300
            }
          ]
        }
      ],
      type: 'CSSValue'
    })
  })

  it('should correctly parse URLs', () => {
    const parser = new Parser()

    expect(
      parser.parse(
        'url(https://www.google.de/request#something?param=true&foo=bar%20)'
      )
    ).toEqual({
      body: [
        {
          type: 'FunctionExpression',
          callee: {
            type: 'Identifier',
            value: 'url'
          },
          params: [
            {
              type: 'URL',
              value:
                'https://www.google.de/request#something?param=true&foo=bar%20'
            }
          ]
        }
      ],
      type: 'CSSValue'
    })
  })

  it('should correctly parse Expressions', () => {
    const parser = new Parser()

    expect(parser.parse('calc(100%+5/2)')).toEqual({
      body: [
        {
          type: 'FunctionExpression',
          callee: {
            type: 'Identifier',
            value: 'calc'
          },
          params: [
            {
              type: 'Expression',
              body: [
                {
                  type: 'Dimension',
                  unit: '%',
                  value: {
                    type: 'Integer',
                    value: 100
                  }
                },
                {
                  type: 'Operator',
                  value: '+'
                },
                {
                  type: 'Integer',
                  value: 5
                },
                {
                  type: 'Operator',
                  value: '/'
                },
                {
                  type: 'Integer',
                  value: 2
                }
              ]
            }
          ]
        }
      ],
      type: 'CSSValue'
    })
  })

  it('should correctly skip whitespaces', () => {
    const parser = new Parser()

    expect(parser.parse('300 400   700')).toEqual({
      body: [
        {
          type: 'Integer',
          value: 300
        },
        {
          type: 'Integer',
          value: 400
        },
        {
          type: 'Integer',
          value: 700
        }
      ],
      type: 'CSSValue'
    })

    expect(
      parser.parse(
        ' 1px   inherit rgba( 255  , 94 ,  0.34 ) , 300ms all linear '
      )
    ).toEqual({
      type: 'MultiValue',
      body: [
        {
          type: 'CSSValue',
          body: [
            {
              type: 'Dimension',
              value: {
                type: 'Integer',
                value: 1
              },
              unit: 'px'
            },
            {
              type: 'Keyword',
              value: 'inherit'
            },
            {
              type: 'FunctionExpression',
              callee: {
                type: 'Identifier',
                value: 'rgba'
              },
              params: [
                {
                  type: 'Integer',
                  value: 255
                },
                {
                  type: 'Integer',
                  value: 94
                },
                {
                  type: 'Float',
                  fractional: {
                    type: 'Integer',
                    value: 34
                  },
                  integer: {
                    type: 'Integer',
                    value: 0
                  }
                }
              ]
            }
          ]
        },
        {
          type: 'CSSValue',
          body: [
            {
              type: 'Dimension',
              unit: 'ms',
              value: {
                type: 'Integer',
                value: 300
              }
            },
            {
              type: 'Identifier',
              value: 'all'
            },
            {
              type: 'Identifier',
              value: 'linear'
            }
          ]
        }
      ]
    })
  })

  // complex test
  it('should return a correct AST', () => {
    const parser = new Parser()

    expect(
      parser.parse('1px inherit rgba(255, 94, 0.34), 300ms all linear')
    ).toEqual({
      type: 'MultiValue',
      body: [
        {
          type: 'CSSValue',
          body: [
            {
              type: 'Dimension',
              value: {
                type: 'Integer',
                value: 1
              },
              unit: 'px'
            },
            {
              type: 'Keyword',
              value: 'inherit'
            },
            {
              type: 'FunctionExpression',
              callee: {
                type: 'Identifier',
                value: 'rgba'
              },
              params: [
                {
                  type: 'Integer',
                  value: 255
                },
                {
                  type: 'Integer',
                  value: 94
                },
                {
                  type: 'Float',
                  fractional: {
                    type: 'Integer',
                    value: 34
                  },
                  integer: {
                    type: 'Integer',
                    value: 0
                  }
                }
              ]
            }
          ]
        },
        {
          type: 'CSSValue',
          body: [
            {
              type: 'Dimension',
              unit: 'ms',
              value: {
                type: 'Integer',
                value: 300
              }
            },
            {
              type: 'Identifier',
              value: 'all'
            },
            {
              type: 'Identifier',
              value: 'linear'
            }
          ]
        }
      ]
    })
  })
})
