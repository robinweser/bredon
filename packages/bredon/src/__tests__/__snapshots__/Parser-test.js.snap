exports[`Parsing CSS values should correctly parse !important 1`] = `
Array [
  "3px !important",
  Object {
    "body": Array [
      Object {
        "body": Array [
          Object {
            "type": "Dimension",
            "unit": "px",
            "value": Object {
              "negative": false,
              "type": "Integer",
              "value": 3,
            },
          },
        ],
        "important": true,
        "multi": false,
        "type": "Value",
      },
    ],
    "type": "ValueList",
  },
]
`;

exports[`Parsing CSS values should correctly parse HexColors 1`] = `
Array [
  "#66FF66",
  Object {
    "body": Array [
      Object {
        "body": Array [
          Object {
            "type": "HexColor",
            "value": "66FF66",
          },
        ],
        "important": false,
        "multi": false,
        "type": "Value",
      },
    ],
    "type": "ValueList",
  },
]
`;

exports[`Parsing CSS values should correctly parse URLs 1`] = `
Array [
  "url(https://www.google.de/request#something?param=true&foo=bar%20)",
  Object {
    "body": Array [
      Object {
        "body": Array [
          Object {
            "callee": "url",
            "params": Array [
              Object {
                "type": "URL",
                "value": "https://www.google.de/request#something?param=true&foo=bar%20",
              },
            ],
            "type": "FunctionExpression",
          },
        ],
        "important": false,
        "multi": false,
        "type": "Value",
      },
    ],
    "type": "ValueList",
  },
]
`;

exports[`Parsing CSS values should correctly parse algebraic signs 1`] = `
Array [
  "-400",
  Object {
    "body": Array [
      Object {
        "body": Array [
          Object {
            "negative": true,
            "type": "Integer",
            "value": 400,
          },
        ],
        "important": false,
        "multi": false,
        "type": "Value",
      },
    ],
    "type": "ValueList",
  },
]
`;

exports[`Parsing CSS values should correctly parse dimensions 1`] = `
Array [
  "300px",
  Object {
    "body": Array [
      Object {
        "body": Array [
          Object {
            "type": "Dimension",
            "unit": "px",
            "value": Object {
              "negative": false,
              "type": "Integer",
              "value": 300,
            },
          },
        ],
        "important": false,
        "multi": false,
        "type": "Value",
      },
    ],
    "type": "ValueList",
  },
]
`;

exports[`Parsing CSS values should correctly parse dimensions 2`] = `
Array [
  "-33.33%",
  Object {
    "body": Array [
      Object {
        "body": Array [
          Object {
            "type": "Dimension",
            "unit": "%",
            "value": Object {
              "fractional": 33,
              "integer": 33,
              "negative": true,
              "type": "Float",
            },
          },
        ],
        "important": false,
        "multi": false,
        "type": "Value",
      },
    ],
    "type": "ValueList",
  },
]
`;

exports[`Parsing CSS values should correctly parse expressions 1`] = `
Array [
  "calc(100%+5/2)",
  Object {
    "body": Array [
      Object {
        "body": Array [
          Object {
            "callee": "calc",
            "params": Array [
              Object {
                "body": Array [
                  Object {
                    "type": "Dimension",
                    "unit": "%",
                    "value": Object {
                      "negative": false,
                      "type": "Integer",
                      "value": 100,
                    },
                  },
                  Object {
                    "type": "Operator",
                    "value": "+",
                  },
                  Object {
                    "negative": false,
                    "type": "Integer",
                    "value": 5,
                  },
                  Object {
                    "type": "Operator",
                    "value": "/",
                  },
                  Object {
                    "negative": false,
                    "type": "Integer",
                    "value": 2,
                  },
                ],
                "type": "Expression",
              },
            ],
            "type": "FunctionExpression",
          },
        ],
        "important": false,
        "multi": false,
        "type": "Value",
      },
    ],
    "type": "ValueList",
  },
]
`;

exports[`Parsing CSS values should correctly parse floats 1`] = `
Array [
  "200.55",
  Object {
    "body": Array [
      Object {
        "body": Array [
          Object {
            "fractional": 55,
            "integer": 200,
            "negative": false,
            "type": "Float",
          },
        ],
        "important": false,
        "multi": false,
        "type": "Value",
      },
    ],
    "type": "ValueList",
  },
]
`;

exports[`Parsing CSS values should correctly parse floats 2`] = `
Array [
  "-.55",
  Object {
    "body": Array [
      Object {
        "body": Array [
          Object {
            "fractional": 55,
            "integer": 0,
            "negative": true,
            "type": "Float",
          },
        ],
        "important": false,
        "multi": false,
        "type": "Value",
      },
    ],
    "type": "ValueList",
  },
]
`;

exports[`Parsing CSS values should correctly parse function expressions 1`] = `
Array [
  "rgba(200,300)",
  Object {
    "body": Array [
      Object {
        "body": Array [
          Object {
            "callee": "rgba",
            "params": Array [
              Object {
                "negative": false,
                "type": "Integer",
                "value": 200,
              },
              Object {
                "negative": false,
                "type": "Integer",
                "value": 300,
              },
            ],
            "type": "FunctionExpression",
          },
        ],
        "important": false,
        "multi": false,
        "type": "Value",
      },
    ],
    "type": "ValueList",
  },
]
`;

exports[`Parsing CSS values should correctly parse identifiers 1`] = `
Array [
  "flex-start",
  Object {
    "body": Array [
      Object {
        "body": Array [
          Object {
            "type": "Identifier",
            "value": "flex-start",
          },
        ],
        "important": false,
        "multi": false,
        "type": "Value",
      },
    ],
    "type": "ValueList",
  },
]
`;

exports[`Parsing CSS values should correctly parse integers 1`] = `
Array [
  "400",
  Object {
    "body": Array [
      Object {
        "body": Array [
          Object {
            "negative": false,
            "type": "Integer",
            "value": 400,
          },
        ],
        "important": false,
        "multi": false,
        "type": "Value",
      },
    ],
    "type": "ValueList",
  },
]
`;

exports[`Parsing CSS values should correctly parse parentheses 1`] = `
Array [
  "(",
  Object {
    "body": Array [
      Object {
        "body": Array [
          Object {
            "type": "Parenthesis",
            "value": "(",
          },
        ],
        "important": false,
        "multi": false,
        "type": "Value",
      },
    ],
    "type": "ValueList",
  },
]
`;

exports[`Parsing CSS values should correctly parse strings 1`] = `
Array [
  "\"hello, it\'s me.\"",
  Object {
    "body": Array [
      Object {
        "body": Array [
          Object {
            "quote": "\"",
            "type": "StringLiteral",
            "value": "hello, it\'s me.",
          },
        ],
        "important": false,
        "multi": false,
        "type": "Value",
      },
    ],
    "type": "ValueList",
  },
]
`;

exports[`Parsing CSS values should correctly skip whitespaces 1`] = `
Array [
  "300 400   700",
  Object {
    "body": Array [
      Object {
        "body": Array [
          Object {
            "negative": false,
            "type": "Integer",
            "value": 300,
          },
          Object {
            "negative": false,
            "type": "Integer",
            "value": 400,
          },
          Object {
            "negative": false,
            "type": "Integer",
            "value": 700,
          },
        ],
        "important": false,
        "multi": true,
        "type": "Value",
      },
    ],
    "type": "ValueList",
  },
]
`;

exports[`Parsing CSS values should correctly skip whitespaces 2`] = `
Array [
  " 1px   inherit rgba( 255  , 94 ,  0.34 ) , 300ms all linear ",
  Object {
    "body": Array [
      Object {
        "body": Array [
          Object {
            "type": "Dimension",
            "unit": "px",
            "value": Object {
              "negative": false,
              "type": "Integer",
              "value": 1,
            },
          },
          Object {
            "type": "Identifier",
            "value": "inherit",
          },
          Object {
            "callee": "rgba",
            "params": Array [
              Object {
                "negative": false,
                "type": "Integer",
                "value": 255,
              },
              Object {
                "negative": false,
                "type": "Integer",
                "value": 94,
              },
              Object {
                "fractional": 34,
                "integer": 0,
                "negative": false,
                "type": "Float",
              },
            ],
            "type": "FunctionExpression",
          },
        ],
        "important": false,
        "multi": true,
        "type": "Value",
      },
      Object {
        "body": Array [
          Object {
            "type": "Dimension",
            "unit": "ms",
            "value": Object {
              "negative": false,
              "type": "Integer",
              "value": 300,
            },
          },
          Object {
            "type": "Identifier",
            "value": "all",
          },
          Object {
            "type": "Identifier",
            "value": "linear",
          },
        ],
        "important": false,
        "multi": true,
        "type": "Value",
      },
    ],
    "type": "ValueList",
  },
]
`;

exports[`Parsing CSS values should return a correct AST 1`] = `
Array [
  "1px inherit rgba(255, 94, 0.34), 300ms all linear",
  Object {
    "body": Array [
      Object {
        "body": Array [
          Object {
            "type": "Dimension",
            "unit": "px",
            "value": Object {
              "negative": false,
              "type": "Integer",
              "value": 1,
            },
          },
          Object {
            "type": "Identifier",
            "value": "inherit",
          },
          Object {
            "callee": "rgba",
            "params": Array [
              Object {
                "negative": false,
                "type": "Integer",
                "value": 255,
              },
              Object {
                "negative": false,
                "type": "Integer",
                "value": 94,
              },
              Object {
                "fractional": 34,
                "integer": 0,
                "negative": false,
                "type": "Float",
              },
            ],
            "type": "FunctionExpression",
          },
        ],
        "important": false,
        "multi": true,
        "type": "Value",
      },
      Object {
        "body": Array [
          Object {
            "type": "Dimension",
            "unit": "ms",
            "value": Object {
              "negative": false,
              "type": "Integer",
              "value": 300,
            },
          },
          Object {
            "type": "Identifier",
            "value": "all",
          },
          Object {
            "type": "Identifier",
            "value": "linear",
          },
        ],
        "important": false,
        "multi": true,
        "type": "Value",
      },
    ],
    "type": "ValueList",
  },
]
`;
